### PR TITLE
Feature/dynamic column gets perf test

### PR DESCRIPTION
**Pull request by davidscottcohen**
Can't perf test dynamic column paging because this branch doesn't have it yet.  But at least basic column-based activities.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/731)

### CODE REVIEW

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import com.palantir.atlasdb.performance.backend.KeyValueServiceConnector;
+
+/**
+ * Performance benchmarks for KVS get with dynamic columns.
+ *
+ * @author coda
+ *
+ */
+@State(Scope.Thread)
+@BenchmarkMode(Mode.AverageTime)
```
Comment (schlosna) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java]: probably want `Mode.SampleTime` to get distribution of times

**Comment by mwakerman** [comment]
Coda, I refactored your tests a little bit to (better) conform to the JMH recommendations (worth taking a quick look at the [samples](http://hg.openjdk.java.net/code-tools/jmh/file/tip/jmh-samples/src/main/java/org/openjdk/jmh/samples/)). Namely, I removed the correctness checking which should be done by unit testing and moved as much of the code to the setup routine as possible (mostly the key generation for the lookup). Worth noting that because the `getAllColumns` benchmark is relatively long-running, I've moved it to using Milliseconds so the numbers are of a similar scale to the other tests.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    public void cleanup() throws Exception {
+        kvs.dropTables(Sets.newHashSet(tableRef1));
+        kvs.close();
+        connector.close();
+    }
+
+    @Benchmark
+    @OutputTimeUnit(TimeUnit.MILLISECONDS)
+    public void getAllColumns(Blackhole bh) {
+        bh.consume(kvs.get(tableRef1, allCells2ReadTimestamp));
```
Comment (schlosna) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java]: you don't need a Blackhole here, just return the Map from get and JMH will consume it for you:

```
public Map<Cell, Value> getAllColumns() {
    return kvs.get(tableRef1, allCells2ReadTimestamp);
}
```

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Benchmark
+    @OutputTimeUnit(TimeUnit.MICROSECONDS)
+    public void getFirstColumnExplicitly(Blackhole bh) {
+        bh.consume(kvs.get(tableRef1, firstCell2ReadTimestamp));
```
Comment (schlosna) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java]: remove blackhole, just return the response from get

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.performance.backend.KeyValueServiceConnector;
+
+/**
+ * Performance benchmarks for KVS get with dynamic columns.
+ *
+ * @author coda
+ *
+ */
+@State(Scope.Thread)
```
Comment (schlosna) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java]: I'm not sure this is what you want as each thread will have its own KVS

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java (hunk 1 above)
Comment (mwakerman) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java]: Cheers David, @rjullman, this Mode is probably worth us looking at when we consider the metrics we actually want to gather. I've just run Coda's tests with `@BenchmarkMode({Mode.AverageTime, Mode.SampleTime})` and this is the (abridged) results:
```
Benchmark                                         (backend)    Mode    Cnt      Score       Error  Units
KvsGetDynamicBenchmarks.getAllColumns              POSTGRES    avgt     20    872.483 ±    57.181  ms/op
KvsGetDynamicBenchmarks.getFirstColumnExplicitly   POSTGRES    avgt     20    285.921 ±    42.203  us/op
KvsGetDynamicBenchmarks.getAllColumns              POSTGRES  sample     40    917.976 ±    42.862  ms/op
KvsGetDynamicBenchmarks.getFirstColumnExplicitly   POSTGRES  sample  69670    290.495 ±    16.659  us/op
```
I'm doing to take a closer look at the [docs for the benchmark modes](http://hg.openjdk.java.net/code-tools/jmh/file/397891feda1b/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_02_BenchmarkModes.java) and we should discuss.

Comment (schlosna) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java]: with sample time you should also get the distribution before the summary:

```
Result "singleGet":
  N = 170167
  mean =   1174.478 ±(99.9%) 79.351 us/op

  Histogram, us/op:
    [      0.000,  125000.000) = 169974
    [ 125000.000,  250000.000) = 177
    [ 250000.000,  375000.000) = 12
    [ 375000.000,  500000.000) = 0
    [ 500000.000,  625000.000) = 0
    [ 625000.000,  750000.000) = 0
    [ 750000.000,  875000.000) = 0
    [ 875000.000, 1000000.000) = 0
    [1000000.000, 1125000.000) = 0
    [1125000.000, 1250000.000) = 0
    [1250000.000, 1375000.000) = 0
    [1375000.000, 1500000.000) = 0
    [1500000.000, 1625000.000) = 0
    [1625000.000, 1750000.000) = 4
    [1750000.000, 1875000.000) = 0

  Percentiles, us/op:
      p(0.0000) =    319.488 us/op
     p(50.0000) =    684.032 us/op
     p(90.0000) =   1452.032 us/op
     p(95.0000) =   1976.320 us/op
     p(99.0000) =   4835.901 us/op
     p(99.9000) = 151257.088 us/op
     p(99.9900) = 238266.876 us/op
     p(99.9990) = 1648766.810 us/op
     p(99.9999) = 1654652.928 us/op
    p(100.0000) = 1654652.928 us/op
```

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java (hunk 2 above)
Comment (mwakerman) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsGetDynamicBenchmarks.java]: You're absolutely right, fixed.

**Comment by joelea** [comment]
This has been failing and untouched for 3 weeks, closing until there's some more movement on it.

**Comment by davidscottcohen** [comment]
Ready to work on this again, here is a new PR

**Pull request merged**